ray-operator/controllers/ray/utils/util.go

```diff
@@ -346,6 +347,10 @@ func PodNotMatchingTemplate(pod corev1.Pod, template corev1.PodTemplateSpec) boo
 					// resource entries do not match
 					return true
 				}
+				if !equality.Semantic.DeepDerivative(container1.VolumeMounts, container2.VolumeMounts) {
```

Comment:
Why do we use `DeepDerivative` rather than `DeepEqual`?

Reply:
DeepEqual causes issues when default volumeMounts are mounted via the Kubernetes cluster. When using DeepEqual, I noticed pod would continuously reconcile.

Reply:
"I also have the problem using reflect.DeepEqual or equality.Semantic.DeepEqual because some fields are set with default non-zero values by some controller (like ImagePullPolicy, RestartPolicy, and so on) if these fields are not set by the operator" - https://github.com/kubernetes-sigs/kubebuilder/issues/592

Reply:
https://github.com/kubernetes-sigs/kubebuilder/issues/592#issuecomment-768113991
https://github.com/kubernetes-sigs/kubebuilder/issues/592#issuecomment-779365927

`DeepDerivative` seems to be risky. Would you mind explaining the behavior of this function and why does it work in our case?

Reply:
Hey @kevin85421, I think those comments are actually what we want. Kubernetes generates empty or nil values, and using DeepEqual will cause problems since it will detect a drift when it shouldn't..

```
DeepDerivative is similar to DeepEqual except that unset fields in a1 are ignored (not compared). This allows us to focus on the fields that matter to the semantic comparison.
```

Reply:
> DeepDerivative is similar to DeepEqual except that unset fields in a1 are ignored (not compared). This allows us to focus on the fields that matter to the semantic comparison.

I knew this. My point is: are there any situations where the worker update could be triggered unintentionally? I have this concern because:

(1) `DeepDeriative` is much more complex than what we want to compare. ([code](https://github.com/kubernetes/apimachinery/blob/master/third_party/forked/golang/reflect/deep_equal.go#L292))
(2) Which fields will be updated by Kubernetes APIServer? Will the changes trigger the worker update unintentionally?

Reply:
Hi Kevin. I believe we have the same concern. Using DeepDeriative for comparing volumeMounts does not trigger unintentionally. 

1. Is is more complex, but it seems that the complexity comes for ensuring fields are properly compared based off data type.
2. The common example is empty volumeMounts and volumes are added via the APIServer causing workers to update unintentionally.